File: python-refactor/Code/RealTimeTraining_PM/02_build_model/RF_01_ranger_LOO_KCJ.R

library(ranger) 
library(tictoc)
rm(list=ls(all=TRUE))

tic() 

target<-c("PM10","PM25")	## Enter the name of dataset (Please use file names with the unified prefix)
type<-c("conc","time","time_conc","cloud")
# pathname<-"//10.72.26.56/irisnas5/GEMS/PM/"  # Nas5/GEMS
pathname<-"/share/irisnas5/GEMS/PM/"  # Nas5/GEMS
setwd(pathname)

num_tree = 500
for (t in 1)
{
  for (i in 1) # Target type
  {
    
    for (yr in 2015:2017)
    {
      if(yr==2015){
        days=365;
      }else{
        days=366;
      }
      
      for (doy in 1:days) #321-323 80
      {
        try({
          cal_list<- Sys.glob((paste("00_EA6km/LOO/",type[t],"/dataset/",target[i],"/",target[i],"_RTT_EA6km_",yr,"_",sprintf("%03d",doy),"*_LOO_*_cal.csv",sep="")))
          val_list<- Sys.glob((paste("00_EA6km/LOO/",type[t],"/dataset/",target[i],"/",target[i],"_RTT_EA6km_",yr,"_",sprintf("%03d",doy),"*_LOO_*_val.csv",sep="")))
          
          for (c in 1:length(cal_list))  #820 985
          {
            tic() 
            
            try({
              cal<-read.csv(file=(cal_list[c])) #For calibration 
              val<-read.csv(file=(val_list[c])) #For validation
              cal<-cal[cal[,24]>0, ]
              val<-val[val[,24]>0, ]
              cal[,24]<-log(cal[,24])
              val[,24]<-log(val[,24])
              
              ## Make Random Forest model using ranger
              t1=Sys.time()
              if (i==1){
                rf_model = ranger(formula=PM10 ~., data=cal, num.trees = num_tree, importance = "permutation", write.forest = TRUE, scale.permutation.importance = TRUE, 
                                  min.node.size = 5, replace = TRUE, splitrule = 'variance', num.threads = 8, save.memory = FALSE, verbose = TRUE)
              } else {
                rf_model = ranger(formula=PM25 ~., data=cal, num.trees = num_tree, importance = "permutation", write.forest = TRUE, scale.permutation.importance = TRUE, 
                                  min.node.size = 5, replace = TRUE, splitrule = 'variance', num.threads = 8, save.memory = FALSE, verbose = TRUE)
              }
              
              
              t2=Sys.time()
              print(t2-t1)
              # importance(rf_model)
              # print('RF training is finished')
              
              # # save variable importance
              # write.table(importance(rf_model),paste("RF/PM/",target[i],"/rf_",name,"_imp_ranger_2.csv",sep=""),sep=",",append=FALSE)
              # 
              # save(rf_model, file=paste("RF/PM/",target[i],"/rf_",name,"_model_ranger.RData",sep=""))
              # print('RF model is saved')
              
              # # Make calibration result
              # pred_cal = predict(rf_model,data=cal)
              # pred_cal<-exp(pred_cal$predictions)
              
              # # save cali prediction
              # write.table(pred_cal, paste("RF/PM/",target[i],"/rf_",name,"_cal_ranger_2.csv",sep=""),sep=",",append=FALSE)
              # print('Predicted cal result is saved')
              
              if(t==3){
                name<-paste(substr(cal_list[c],37,70),sep="")
              }else if (t==4){
                name<-paste(substr(cal_list[c],33,66),sep="")
              }else{
                name<-paste(substr(cal_list[c],32,65),sep="")
              }
              
              # Validation result
              pred_val = predict(rf_model,data=val)
              pred_val<-exp(pred_val$predictions)
              
              # save vali prediction
              write.table(pred_val, paste("00_EA6km/LOO/",type[t],"/RF/",target[i],"/rf_",name,"_val_ranger.csv",sep=""),sep=",",append=FALSE)
              # print('Predicted val result is saved')
              
              
              toc()      
            },silent = TRUE)
          } # LOO_list
        },silent = TRUE)
      } # doy
    } #yr
  } # tatget
} #type
toc()

File: python-refactor/Code/RealTimeTraining_PM/02_build_model/RF_01_ranger_LOO_KCJ.py
### Package Import
import sys
import os
base_dir = os.environ['GEMS_HOME']
project_path = os.path.join(base_dir, 'python-refactor')
sys.path.insert(0, project_path)
from Code.utils import matlab

import numpy as np
import glob
import pandas as pd
#from thundergbm import TGBMClassifier
from sklearn.ensemble import RandomForestRegressor

### Setting path
data_base_dir = os.path.join('/data2', 'sehyun', 'Data')
path_grid_raw = os.path.join(data_base_dir, 'Raw', 'grid')
path_ea_goci = os.path.join(data_base_dir, 'Preprocessed_raw', 'EA_GOCI6km')
path_rtt = os.path.join(data_base_dir, 'Preprocessed_raw', 'RTT')
path_loo = os.path.join(data_base_dir, 'Preprocessed_raw', 'LOO')

target = ["PM10", "PM25"] ## Enter the name of dataset (Please use file names with the unified prefix)
type_list = ["conc","time","time_conc","cloud"]
# pathname<-"//10.72.26.56/irisnas5/GEMS/PM/"  # Nas5/GEMS

num_tree = 500
YEARS = [2016]
for t in [0]: # type
    for i in [0]: # Target 
        for yr in YEARS:
            if yr%4==0: days = 365
            else: days = 366
            
            for doy in range(1, days+1): # 321-323 80
                try:
                    fname = f'{target[i]}_RTT_EA6km_{yr}_{doy:03d}*_LOO_*_cal.csv'
                    cal_list = glob.glob(os.path.join(path_roo,type_list[t],"/dataset/",target[i],fname))
                    fname = f'{target[i]}_RTT_EA6km_{yr}_{doy:03d}*_LOO_*_val.csv'
                    val_list = glob.glob(os.path.join(path_roo,type_list[t],"/dataset/",target[i],fname))
                    
                    for c in range(len(cal_list)): # 820 985
                        try:
                            cal = pd.read_csv(cal_list[c])
                            val = pd.read_csv(val_list[c])

                            cal = cal.loc[cal.iloc[:, 23]>0, :]
                            val = val.loc[val.iloc[:, 23]>0, :]
                            cal.iloc[:, 23] = np.log(cal.iloc[:, 23])
                            val.iloc[:, 23] = np.log(val.iloc[:, 23])
                            
                            ## Mask Random Forest model using ranger
                            ti = time.time()
                            params = {
                                    n_estimators:num_tree,
                                    criterion:'mse', # splitrule = 'variance', “mse” for the mean squared error, which is equal to variance reduction, The function to measure the quality of a split.
                                    min_samples_leaf:5, # min.node.size = 5, 
                                    n_jobs:8, # num.threads = 8, 
                                    verbose:2, # verbose = TRUE. for debugging. 
                                    bootstrap:True, # replace = TRUE, excatly same? not sure...
                                    max_depth:None, # max.depth=default 0. unlimited depth
                                    min_samples_split:1, # Fraction of observations to sample. Default is 1 for sampling with replacement
                                    #importance = "permutation", 
                                    #write.forest = TRUE, 
                                    #scale.permutation.importance = TRUE, 
                                    #save.memory = FALSE, 
                            }
                            rf_model = RandomForestRegressor(**params)
                            rf_model.fit(cal)
                            t2 = time.time()
                            print (t2-t1)
                            
                            if t==3:
                                name = os.path.basenanme(cal_list[c])[36:70]
                            elif t==4:
                                name = os.path.basenanme(cal_list[c])[32:66]
                            else:
                                name = os.path.basenanme(cal_list[c])[31:65]
                            
                            # Validation reslut
                            pred_val = rf_model.predict(val.drop([target[i]]), val[target[i]])
                            pred_val = np.exp(pre_val)
                            
                            # save vali pred
                            fname = f'rf_{name}_val_ranger.csv'
                            pred_val = pd.DataFrame(pred_val)
                            matlab.check_make_dir(os.path.join(paht_loo, type_list[t],"/RF/",target[i]))
                            pred_val.to_csv(os.path.join(paht_loo, type_list[t],"/RF/",target[i], fname), sep=",")
                            print('Predicted val result is saved')
                        except:
                            pass
                    # LOO list
                    print (doy)
                except:
                    pass
                # doy
                print (yr)
            # yr
            print (target[i])
        # target
        print (type_list[t])
    # type                          